staging/src/k8s.io/cli-runtime/pkg/genericclioptions/resource/visitor.go

```diff
@@ -452,7 +457,10 @@ func ExpandPathsToFileVisitors(mapper *mapper, paths string, recursive bool, ext
 		if err != nil {
 			return err
 		}
-
+		if isKustomizationDir(path) {
```

Comment:
I don't think this is backwards compatible.  Anyone using `kubectl create -f DIR` will see something different happen after this change lands, which means existing CLI workflows could break.

Reply:
I just tested the behaviour.  `kubectl create` and `kubectl apply` both fail with a kustomization.yaml present:

```
error: error validating ".../k8s.io/examples/guestbook-go/kustomization.yaml": error validating data: [apiVersion not set, kind not set]; if you choose to ignore these errors, turn validation off with --validate=false
```

It looks like we have a separate issue, which is that we don't validate all files before starting to apply, which is contrary to what I would have expected.

But I don't think users can be using `kubectl create -f DIR` or `kubectl apply -f DIR` today with a dir containing kustomization.yaml.

Reply:
@smarterclayton The existing kubectl doesn't work with directories with a kustomization.yaml as @justinsb explained. With this PR, kubectl will be able to recognize a directory with a kustomization.yaml. For any directories without kustomization.yaml, there is no change in kubectl's behavior.

Reply:
If someone has validate=false off, what happens?  If it also fails, then my primary concern is addressed.

Reply:
With validate=false, it fails with similar error
```
error: unable to decode "kustomization.yaml": Object 'Kind' is missing in `<truncated>`
```